test/reporting_api_parser_test.dart

```diff
@@ -0,0 +1,206 @@
+        ''';
+
+      final actual = ReportingQueryParser.parseQueryIdFromJsonString(input);
+
+      final expected = '1234567';
+      expect(actual, expected);
+    });
+
+    test('parses download path from a json string', () {
+      input = '''
+        {
+           "kind": "doubleclickbidmanager#query",
+           "queryId": "1234567",
+            "metadata": {
+               "title": "jin-test",
+               "dataRange": "LAST_14_DAYS",
+               "format": "CSV",
+               "running": true,
+               "googleCloudStoragePathForLatestReport": "I-am-a-download-link",
+               "latestReportRunTimeMs": "1357924",
+               "sendNotification": false
+            },
+          "params": {
+            "type": "TYPE_GENERAL",
+            "groupBys": [
+               "FILTER_INSERTION_ORDER",
+               "FILTER_DATE"
+            ],
+            "filters": [
+             {
+                "type": "FILTER_ADVERTISER",
+                "value": "164337"
+              }
+            ],
+            "metrics": [
+              "METRIC_REVENUE_USD"
+            ]
+          },
+          "schedule": {
+            "frequency": "DAILY"
+          }
+        }
+        ''';
+
+      final expected =
+          ReportingQueryParser.parseDownloadPathFromJsonString(input);
+
+      final actual = 'I-am-a-download-link';
+      expect(actual, expected);
+    });
+
+    group('parses revenue from:', () {
+      String generateInput(String reportBody) {
+        return '''
+        {
+          "gapiRequest":{
+            "data":
+            {
+              "body":"$reportBody",
+              "headers":
+              {
+                "cache-control":"",
+                "content-disposition":"",
+                "content-length":"",
+                "content-type":"text/csv"
+              },
+              "status":200
+            }
+          }
+        }
+        ''';
+      }
+
+      test('a report json string that contains no revenue values', () {
+        input = generateInput('Insertion Order ID, Revenue');
+
+        final actual = ReportingQueryParser.parseRevenueFromJsonString(input);
+
+        expect(actual is Multimap<String, InsertionOrderDailySpend>, true);
+        expect(actual.isEmpty, true);
+      });
+
+      test('a report json string that contains a single row of revenue value',
+          () {
+        final reportBody = 'Insertion Order ID, Date, Revenue, Impression\\n'
+            '123456,2020/01/01,88.88,1000\\n';
+        input = generateInput(reportBody);
+
+        final actual = ReportingQueryParser.parseRevenueFromJsonString(input);
+
+        final expectedSpending = InsertionOrderDailySpend((b) => b
+          ..date = DateTime(2020, 1, 1)
+          ..revenue = '88.88'
+          ..impression = '1000');
+
+        expect(actual is Multimap<String, InsertionOrderDailySpend>, true);
+        expect(actual.length, 1);
+        expect(actual['123456'].length, 1);
```

Comment:
143 will fail if this is not true

Reply:
Oooo, 142 and 143 seems redundant, seems like checking the length here is not necessary?

Reply:
I copied the mapEqual function over and make it work for multiMap as well. No longer doing these checks now.